editor/plugins/path_3d_editor_plugin.cpp

```diff
@@ -40,17 +40,28 @@
 #include "scene/gui/menu_button.h"
 #include "scene/resources/curve.h"
 
-static bool _is_in_handle(int p_id, int p_num_points) {
-	int t = (p_id + 1) % 2;
-	int idx = (p_id + 1) / 2;
-	// order of points is [out_0, out_1, in_1, out_2, in_2, ... out_n-1, in_n-1, in_n]
-	if (idx == 0) {
-		return false;
-	} else if (idx == (p_num_points - 1)) {
-		return true;
-	} else {
-		return (t == 1);
-	}
+// Map handle id to control point id and handle type.
+enum HandleType {
+	IN = 0,
+	OUT,
+	TILT,
+};
+
+struct HandleInfo {
+	int point_idx; // Index of control point.
+	HandleType type; // Type of this handle.
+};
+
+static HandleInfo _handle_to_point(int p_id) {
+	HandleInfo info;
+	info.point_idx = p_id / 3;
+	info.type = HandleType(p_id % 3);
+	return info;
+}
+
+static int _point_to_handle(HandleInfo p_info) {
+	int idx = p_info.point_idx * 3 + int(p_info.type);
+	return idx;
```

Comment:
It is redundant to convert the HandleType by index even though the struct has the HandleType. Conversion to arrays makes sense in cases like storing `.tres` where structs cannot be used, but that is not the case here.

Reply:
Hi thanks for the review! But I didn't quite get what you mean. Could you elaborate what does "Conversion to arrays" mean? It seems there is not any "arrays" in this piece of code.

Reply:
```
Vector<Vector3> handle_lines;
Vector<Vector3> primary_handle_points;
Vector<Vector3> secondary_handle_points;
PackedInt32Array _secondary_handle_ids; // Avoid shadowing member on Node3DEditorGizmo
```
I am meaning something in these. Perhaps if you use struct like `Vector<HandleInfo>` would avoid the unnecessary identification of the HandleType by index.

Reply:
Thanks for the clarification.

> It is redundant to convert the HandleType by index even though the struct has the HandleType
In this PR there are two ways to identify a handle
- as an unique `int` for each handle without regards to its type
- as curve point index it attached to, and the type(in, out, tilt)

The first one is required by the `void EditorNode3DGizmo::add_handles()` API, but difficult to work with. So I convert it into the second(represented by a `HandleInfo`), work with it, and convert it back to the first one before call `add_handles`. The two helper functions `_point_to_handle`, `_handle_to_point` do this job, so they are necessary.

> Perhaps if you use struct like Vector<HandleInfo> would avoid the unnecessary identification of the HandleType by index.

These arrays are just local temporary containers used in collecting positions for drawing handles, and . Due to the API design of `void EditorNode3DGizmo::add_handles()`,  primary handles and secondary handles have to be drawn in two pass. So it is necessary to separate into different arrays. See their use in the following code.

```
		if (handle_lines.size() > 1) {
			add_lines(handle_lines, path_thin_material);
		}
		if (primary_handle_points.size()) {
			add_handles(primary_handle_points, handles_material);
		}
		if (secondary_handle_points.size()) {
			add_handles(secondary_handle_points, sec_handles_material, _secondary_handle_ids, false, true);
		}
```

>  it is better to use LocalVector<T> for the performance.

They reason I use Vector<T> instead of LocalVector<T> is also due to the API. 
```
EditorNode3DGizmo::add_handles(const PackedVector3Array &p_handles, const Ref<Material> &p_material, const PackedInt32Array &p_ids = Vector<int>(), bool p_billboard = false, bool p_secondary = false)
```
`add_handles` method takes `PackedVector`(Alias for `Vector<T>`) as parameter. If we use `LocalVector<T>` as collecting containers, then there is a type conversion from `LocalVector<T>` to `Vector<T>`, which is expensive because it involves memory copy.  See its implementation
```
	operator Vector<T>() const {
		Vector<T> ret;
		ret.resize(size());
		T *w = ret.ptrw();
		memcpy(w, data, sizeof(T) * count);
		return ret;
	}
```

Reply:
The change to LocalVector is not required, but something about the former explanation seems not good way to implement it. The conversion using indexes is clearly a hack and should be changed to a more straightforward/clear design.

Reply:
> conversion between index and type is obvious a hack 

That's true

> and must be changed to a more straightforward/clear design.

but that's again due to the limitation of `EditorNode3DGizmo::add_handles()`. Let's say we have a curve of 8 control points. Then there are 3 handles(in, out, tilt) for each control point, or 24 handles in total. For the `EditorNode3DGizmo::add_handles()` they are referred by `int` value 0-23.

Then in my gizmo code, this integer is not very meaningful. We always need to devise a scheme to find out which control point the handle is attached to, and what type this handle is. A reasonable scheme is the division and modulus of 3.  For example handle number 13 is attached to control point 13 / 3 = 3, and is of type 13 % 3 = 1.

That is basically how the old code did it(but in a messier way), compare the old way

```
// (p_id + 1) Accounts for the first point only having an "out" handle
	int idx = (p_id + 1) / 2;

static bool _is_in_handle(int p_id, int p_num_points) {
	int t = (p_id + 1) % 2;
	int idx = (p_id + 1) / 2;
	// order of points is [out_0, out_1, in_1, out_2, in_2, ... out_n-1, in_n-1, in_n]
	if (idx == 0) {
		return false;
	} else if (idx == (p_num_points - 1)) {
		return true;
	} else {
		return (t == 1);
	}
```
and the new way
```
static HandleInfo _handle_to_point(int p_id) {
	HandleInfo info;
	info.point_idx = p_id / 3;
	info.type = HandleType(p_id % 3);
	return info;
}

static int _point_to_handle(HandleInfo p_info) {
	int idx = p_info.point_idx * 3 + int(p_info.type);
	return idx;
```
I'd say it's an improvement. Of course if you have a better way I am all ears.

Reply:
You just create a map like `HashMap<int id, HandleInfo handle>` with the id of the handle as the key. More simply, if the id is consistent, then an array would be sufficient instead of a map.

Reply:
How is the `HashMap` improve the code? Let's say I add a `HashMap`

```
HashMap<int, HandleInfo> handle_map;
```
then I still need to populate it with
```
for(int id=0; i < 24; i++) {
    handle_map.add(i, _handle_to_point(i));
```
then we I need to find out the information of handle 13, I call
```
   HandleInfo info = handle_map[13];
``` 

But this HashMap is completely redundant, I can simply call
```
   HandleInfo info = _handle_to_point(13);
```
cache data that can be computed  on the fly not only takes more code,  more memory, and introduce potential synchronization problem. It is only needed when performance is an issue. Here test build shows that on-the-fly computation is fast enough.

Reply:
The type decision by id in `_is_in_handle()` is redundant, so it is necessary to get it from type in struct.

Reply:
> The type decision by id is definitely redundant
It is not. `_is_in_handle` is from the old code before this PR. the first two lines are
```
static bool _is_in_handle(int p_id, int p_num_points) {
	int t = (p_id + 1) % 2;
	int idx = (p_id + 1) / 2;
```
it basically did exactly what I am doing. `t` for handle type, 0 is in, 1 is out. `idx` for the control point idx.

Reply:
In essence, what you are doing with `_handle_to_point()` is just needlessly adding more structs than before.

For example, you should consider like the following when adding a handle:

```cpp
add_handles(vec, mat, id); // In 0
handles_info[id].type = HANDLE_TYPE_IN_0;

add_handles(vec, mat, id + 1); // In 1
handles_info[id + 1].type = HANDLE_TYPE_IN_0;

add_handles(vec, mat, id + 2); // Out 0
handles_info[id + 2].type = HANDLE_TYPE_IN_0;

add_handles(vec, mat, id + 3); // Out 1
handles_info[id + 3].type = HANDLE_TYPE_IN_0;
```

This way, if you want to know the type of the handle later, you don't have to convert the id.

Reply:
`add_handles` doesn't add handle one by one, it add handles in batch. So I need to collect all handles in two arrays(primary and secondary), and add them with two calls. 

If I understand correctly, you mean some code like this need to be used
```
Vector<HandleInfo> handles_info;

// cp_id for control point id
for(int cp_id=0; cp_id<cp_size; cp_id++) {
    handles_info[id*3+0] = HandleType::IN;
    handles_info[id*3+1] = HandleType::OUT;
    handles_info[id*3+2] = HandleType::TILT;
```
and get handle type by
```
handles_info[n].type;
```

but this is still just cache of what can be computed on-the-fly by
```
_handle_to_point(n).type;
```
subjecting to the same problems of cache.

Reply:
Yes, using `handles_info[n].type;` instead of `_handle_to_point(n).type;` should be a much cleaner implementation.

Reply:
done